cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -190,6 +234,70 @@ inline void ivfpq_encode(uint32_t pq_dim,
   }
 }
 
+template <uint32_t BlockDim, typename IdxT>
+__launch_bounds__(BlockDim) __global__ void ivfpq_fill_new_data_kernel(
+  device_mdspan<pq_vec_t, pq_int_vec_exts, row_major> pq_dataset,
+  device_mdspan<IdxT, extent_1d<IdxT>, row_major> pq_indices,
+  device_mdspan<IdxT, extent_1d<uint32_t>, row_major> pq_cluster_offsets,
+  device_mdspan<uint32_t, extent_1d<uint32_t>, row_major> pq_cluster_sizes,
+  device_mdspan<const pq_vec_t, pq_new_vec_exts, row_major> new_codes,
+  const IdxT* new_data_offsets,
+  const IdxT* new_data_indices,
+  const uint32_t* new_data_labels,
+  std::variant<IdxT, const IdxT*> src_offset_or_indices)
+{
+  size_t new_codes_ix = size_t(threadIdx.x) + BlockDim * size_t(blockIdx.x);
+  if (new_codes_ix >= new_codes.extent(0)) return;
+  auto source_ix = new_data_indices[new_codes_ix];
+  auto l         = new_data_labels[source_ix];
+
+  // Get the offset into the extended dataset
+  auto in_cluster_ix = new_codes_ix - new_data_offsets[l];
+  auto db_offset     = pq_cluster_offsets(l) + pq_cluster_sizes(l) + in_cluster_ix;
+
+  // copy db indices
+  pq_indices(db_offset) = std::holds_alternative<IdxT>(src_offset_or_indices)
+                            ? std::get<IdxT>(src_offset_or_indices) + source_ix
+                            : std::get<const IdxT*>(src_offset_or_indices)[source_ix];
+
+  // copy db codes
+  using group    = Pow2<pq_dataset.static_extent(2)>;
+  auto group_ix  = group::div(db_offset);
+  auto member_ix = group::mod(db_offset);
+  for (uint32_t chunk_ix = 0; chunk_ix < new_codes.extent(1); chunk_ix++) {
+    pq_dataset(group_ix, chunk_ix, member_ix) = new_codes(new_codes_ix, chunk_ix);
+  }
+}
+
+template <typename IdxT>
+inline void ivfpq_fill_new_data(
```

Comment:
Please add docstring explaining how the new data is filled.

Reply:
[Outdated code]